cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -2285,6 +2301,11 @@ Result<FileInfo> S3FileSystem::GetFileInfo(const std::string& s) {
 
     auto outcome = impl_->client_->HeadObject(req);
     if (outcome.IsSuccess()) {
+      bool ends_in_slash = s[s.size() - 1] == '/';
+      if (outcome.GetResult().GetContentLength() == 0 && ends_in_slash) {
+        info.set_type(FileType::Directory);
+        return info;
+      }
```

Comment:
It's not, I've removed it.  I think I was just trying to make sense of the way `GetFileInfo` works with directories while debugging failures.  We create these kinds of files when we create directories recursively.  For example, when we create `foo/bar/baz` recursively (in some bucket `my_bucket`) we will create empty files `foo/` and `foo/bar/` and I was concerned we were reading these as files.

However, if `s` is `foo/bar/` here then the trailing slash will be removed in `S3Path::FromString`.  So this newly added branch is impossible to encounter.  Instead, we will look for `foo/bar` and fail to find it.  We will then eventually decide it is a directory because we search for `foo/bar/` as part of `IsNonEmptyDirectory`.